### PR TITLE
OMTF o2o upgrade

### PR DESCRIPTION
**Pull request by michalszleper**
#### PR description:

Created a new OMTF configuration object L1TMuonOverlapFwVersion to be used in o2o, a dedicated ES producer and a new o2o online producer.  Also necessary was an appropriate modification in L1TMuonOverlapObjectKeysOnlineProd.cc and little additions and modifications in several other packages in order to include the new record.  There is also some standalone test code of the new record in Utilities, which is of lesser importance.

#### PR validation:

The code has been tested with o2o.sh independently by myself and Hyejin Kwon, with input from the OMDS and output in a local sqlite file.  Everything was checked to work such as expected from o2o.  Writing and reading the new record in a standalone way was also tested using the code in Utilities.

### CODE REVIEW

**Comment by tvami** [comment]
Hi @michalszleper please note that you have forgotten to do `code-checks` so the PR tests cant be triggered.

File: CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py [hunk 1]
```diff
@@ -41,6 +41,12 @@ def initL1SubsystemsExt( tagBaseVec = [],
             type = cms.string('L1TMuonEndCapForest'),
             key = cms.string(objectKey)
         ),
+        cms.PSet(
+            record = cms.string('L1TMuonOverlapFwVersionO2ORcd'),
+            tag = cms.string('L1TMuonOverlapFwVersion_' + tagBaseVec[ L1CondEnumExt.L1TMuonOverlapFwVersion ]),
+            type = cms.string('L1TMuonOverlapFwVersion'),
+            key = cms.string(objectKey)
+        ),
```
Comment (tvami) [inline CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py:49]: Why is this not just read from the GT? This kind of direct access to tags is only allowed in tests files, this is not a test file, right?

File: L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+class L1TMuonOverlapFwVersionOnlineProd : public L1ConfigOnlineProdBaseExt<L1TMuonOverlapFwVersionO2ORcd, L1TMuonOverlapFwVersion> {
+private:
+  const bool transactionSafe;
+  const edm::ESGetToken<L1TMuonOverlapFwVersion, L1TMuonOverlapFwVersionRcd> baseSettings_token;
+
+public:
+  std::unique_ptr<const L1TMuonOverlapFwVersion> newObject(const std::string& objectKey, const L1TMuonOverlapFwVersionO2ORcd& record) override;
+
+  L1TMuonOverlapFwVersionOnlineProd(const edm::ParameterSet&);
+  ~L1TMuonOverlapFwVersionOnlineProd(void) override {}
```
Comment (tvami) [inline L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:24]:
```suggestion
  ~L1TMuonOverlapFwVersionOnlineProd(void) override = default;
```
This is the new suggestion from the framework team. Please apply it elsewhere too when this comes up

File: L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+L1TMuonOverlapFwVersionOnlineProd::L1TMuonOverlapFwVersionOnlineProd(const edm::ParameterSet& iConfig)
+    : L1ConfigOnlineProdBaseExt<L1TMuonOverlapFwVersionO2ORcd, L1TMuonOverlapFwVersion>(iConfig),
+    transactionSafe(iConfig.getParameter<bool>("transactionSafe")),
+    baseSettings_token(wrappedSetWhatProduced(iConfig).consumes()) {}
+
+std::unique_ptr<const L1TMuonOverlapFwVersion> L1TMuonOverlapFwVersionOnlineProd::newObject(
+    const std::string& objectKey, const L1TMuonOverlapFwVersionO2ORcd& record) {
+
+  const L1TMuonOverlapFwVersionRcd& baseRcd = record.template getRecord<L1TMuonOverlapFwVersionRcd>();
+  auto const& baseSettings = baseRcd.get(baseSettings_token);
```
Comment (tvami) [inline L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:57]: ES get() is deprecated, move to the new method with the tokens and `getData` / `getHandle`. There is 4 occurance of `.get(`, so you need to resolve all of those

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+void L1TMuonOverlapFwVersionTester::analyze(const edm::Event &iEvent, const edm::EventSetup &evSetup) {
+  edm::ESHandle<L1TMuonOverlapFwVersion> handle1;
+  if (isO2Opayload)
+    evSetup.get<L1TMuonOverlapFwVersionO2ORcd>().get(handle1);
+  else
+    evSetup.get<L1TMuonOverlapFwVersionRcd>().get(handle1);
+  std::shared_ptr<L1TMuonOverlapFwVersion> ptr1(new L1TMuonOverlapFwVersion(*(handle1.product())));
+
+  cout << "Contents of L1TMuonOverlapFwVersion: " << endl;
```
Comment (tvami) [inline L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc:44]: `cout` is not thread safe, move to MsgLogger's LogInfo for example

File: CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py (hunk 1 above)
Comment (hjkwon260) [reply CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py:49]: Hi @tvami we directly fetch/update desired tags via O2O process, afaik GT is not involved in the L1 O2O process

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  cout << " layersVersion() = " << ptr1->layersVersion() << endl;
+  cout << " patternsVersion() = " << ptr1->patternsVersion() << endl;
+  cout << " synthDate() = " << ptr1->synthDate() << endl;
+
+  if (writeToDB) {
+    edm::Service<cond::service::PoolDBOutputService> poolDb;
+    if (poolDb.isAvailable()) {
+      cout << "Writing payload to DB" << endl;
+      cond::Time_t firstSinceTime = poolDb->beginOfTime();
+      poolDb->writeOne(
```
Comment (tvami) [inline L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc:56]: `writeOne` is deprecated, please use the new method

File: CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py (hunk 1 above)
Comment (tvami) [reply CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py:49]: Hi @ggovi this is the line we are discussing now in the AlCa meeting

Comment (ggovi) [reply CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py:49]: @tvami thanks! Is it possible to get the tag for hlt with something like "autoCond['run2_data']" ? If this is possibile, I agree it is the way to go.

Comment (tvami) [reply CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py:49]: Yes, we have that key:
https://github.com/cms-sw/cmssw/blob/master/Configuration/AlCa/python/autoCond.py#L35

Comment (ggovi) [reply CondTools/L1TriggerExt/python/L1SubsystemParamsExt_cfi.py:49]: Ok, I would propose to post-pone to a future PR the transition to the tag resolution from GT. We may need a dedicate python function for this scope.

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  ~L1TMuonOverlapFwVersionTester(void) override {}
+};
+
+void L1TMuonOverlapFwVersionTester::analyze(const edm::Event &iEvent, const edm::EventSetup &evSetup) {
+  edm::ESHandle<L1TMuonOverlapFwVersion> handle1;
+  if (isO2Opayload)
+    evSetup.get<L1TMuonOverlapFwVersionO2ORcd>().get(handle1);
+  else
+    evSetup.get<L1TMuonOverlapFwVersionRcd>().get(handle1);
+  std::shared_ptr<L1TMuonOverlapFwVersion> ptr1(new L1TMuonOverlapFwVersion(*(handle1.product())));
```
Comment (ggovi) [inline L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc:42]: It looks like you have an additional copy that could be removed, for example by coding:
const L1TMuonOverlapFwVersion& obj = *handle1.product();
In this way, you can move the below call to writeOne to writeOneIOv(obj

File: L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc (hunk 2 above)
Comment (tvami) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:26]: @michalszleper you havent changed this yet

**Comment by tvami** [comment]
Hi @michalszleper thanks for doing the code-checks, please address the rest of the comments as well

File: L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc (hunk 3 above)
Comment (michalszleper) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:58]: Not sure what is wrong here.  I am using a token to get data from the record, like suggested in
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_a_Record
Is this not allowed anymore?

Comment (tvami) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:58]: Ah yes, you are right, you are getting this from the record and not the ES. In this case this is fine, sorry for the noise!

Comment (tvami) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:58]: Feel free to mark this comment "resolved"

Comment (tvami) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:58]: On the other hand there are cases when you do take it from the ES and thus it's wrong, let me point to another example

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  }
+  ~L1TMuonOverlapFwVersionTester(void) override {}
+};
+
+void L1TMuonOverlapFwVersionTester::analyze(const edm::Event &iEvent, const edm::EventSetup &evSetup) {
+  edm::ESHandle<L1TMuonOverlapFwVersion> handle1;
+  if (isO2Opayload)
+    evSetup.get<L1TMuonOverlapFwVersionO2ORcd>().get(handle1);
+  else
+    evSetup.get<L1TMuonOverlapFwVersionRcd>().get(handle1);
```
Comment (tvami) [inline L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc:41]: please follow this
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc (hunk 6 above)
Comment (tvami) [reply L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc:42]: also I think the handle1 is not used for anything else then just to get the product of it, so one can use the getData() method of the ES, which currently is using a get() method and the handle. I pointed to the twiki to follow in https://github.com/cms-sw/cmssw/pull/35876/files/250d619bf5e78fa43e2b76a77b872ab6993d3d64#r747562872

File: L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc (hunk 3 above)
Comment (tvami) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc:58]: Here https://github.com/cms-sw/cmssw/pull/35876/files/250d619bf5e78fa43e2b76a77b872ab6993d3d64#r747562872

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
Hi @michalszleper please fix this error
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5e7108/20478/llvm-analysis/legacy-mod-sa.txt
as well, thanks!

File: L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc (hunk 2 above)
Comment (michalszleper) [reply L1TriggerConfig/L1TConfigProducers/src/L1TMuonOverlapFwVersionOnlineProd.cc]: I changed to override = default, but only in the files which are part of this PR.

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc (hunk 4 above)
Comment (michalszleper) [reply L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc]: Done.

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc (hunk 5 above)
Comment (michalszleper) [reply L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc]: Done.

File: L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc (hunk 7 above)
Comment (michalszleper) [reply L1TriggerConfig/Utilities/src/L1TMuonOverlapFwVersionTester.cc]: Done.

**Comment by michalszleper** [comment]
I changed edm::EDAnalyzer to edm::one::EDAnalyzer, hope that's OK.
________________________________________

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
Ok it looks good to me, I dont have any further comments, @yuanchao anything from your side?

**Comment by yuanchao** [comment]
Regardless to set GT via autoCond, I see most problems are from the deprecated ES:get and edm::EDAnalyzer are actually from the included L1T XXXviewers and XXXwriters. They need to be fixed in separated PRs. For others, I don't have further comments. It's ok with me.

**Comment by tvami** [comment]
+alca

**Comment by ggovi** [comment]
+db

**Comment by rekovic** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**